### PR TITLE
feat: use a regex to fetch SPICE files for default Environments

### PR DESCRIPTION
**Pull request by kyle-cochran**
There might be a better place for this logic.

### CODE REVIEW

File: src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp [hunk 1]
```diff
@@ -228,16 +231,35 @@ Engine::Mode Engine::DefaultMode()
...
+        }
+        return result;
+    };
+
     static const Array<Kernel> defaultKernels = {
 
-        Kernel::File(File::Path(aLocalRepository.getPath() + Path::Parse("naif0012.tls"))),  // Leap seconds
-        Kernel::File(File::Path(aLocalRepository.getPath() + Path::Parse("de430.bsp"))),     // Ephemeris
-        Kernel::File(File::Path(aLocalRepository.getPath() + Path::Parse("pck00011.tpc"))
+        Kernel::File(File::Path(findKernels(std::regex("naif[0-9]*.tls")).accessFirst())),  // Leap seconds
```
Comment (lucas-bremond) [inline src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp:256]: Isn't `accessFirst` going to segfault if no file exists?

File: src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp [hunk 2]
```diff
@@ -228,16 +231,35 @@ Engine::Mode Engine::DefaultMode()
...
+        Kernel::File(File::Path(aLocalRepository.getPath() + Path::Parse("de430.bsp"))),    // Ephemeris
+        Kernel::File(File::Path(aLocalRepository.getPath() + Path::Parse("pck00010.tpc"))
```
Comment (lucas-bremond) [inline src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp:258]: Why reverting to an older version? (I was looking at the files [here](https://naif.jpl.nasa.gov/pub/naif/generic_kernels/pck/))

File: test/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE.test.cpp [hunk 3]
```diff
@@ -339,3 +341,25 @@ TEST(OpenSpaceToolkit_Physics_Environment_Ephemerides_SPICE, AutomaticMode)
...
+    using ostk::core::fs::Directory;
+
+    using ostk::physics::env::ephem::spice::Engine;
+    using ostk::physics::env::ephem::spice::Kernel;
+
+    {
+        Directory spiceDir =
+            Directory::Path(Path::Parse("/app/test/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE"));
+
+        Array<Kernel> kernels = Engine::DefaultKernels(spiceDir);
```
Comment (lucas-bremond) [inline test/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE.test.cpp:358]:
```suggestion
        const Array<Kernel> kernels = Engine::DefaultKernels(spiceDir);
```

File: test/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE.test.cpp [hunk 4]
```diff
@@ -339,3 +341,25 @@ TEST(OpenSpaceToolkit_Physics_Environment_Ephemerides_SPICE, AutomaticMode)
...
+{
+    using ostk::core::ctnr::Array;
+    using ostk::core::fs::Path;
+    using ostk::core::fs::Directory;
+
+    using ostk::physics::env::ephem::spice::Engine;
+    using ostk::physics::env::ephem::spice::Kernel;
+
+    {
+        Directory spiceDir =
```
Comment (lucas-bremond) [inline test/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE.test.cpp:355]:
```suggestion
        const Directory spiceDirectory =
```

File: src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp (hunk 2 above)
Comment (kyle-cochran) [reply src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp:258]: Good catch. This is a testing artifact.

File: src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp (hunk 1 above)
Comment (kyle-cochran) [reply src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp]: Added throw to raise a sensible error if nothing found.

File: src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp (hunk 2 above)
Comment (kyle-cochran) [reply src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp]: Went ahead and make this "pck[0-9]*.tpc"

File: test/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE.test.cpp [hunk 5]
```diff
@@ -339,3 +341,28 @@ TEST(OpenSpaceToolkit_Physics_Environment_Ephemerides_SPICE, AutomaticMode)
...
+
+    using ostk::physics::env::ephem::spice::Engine;
+    using ostk::physics::env::ephem::spice::Kernel;
+
+    {
+
+        Directory spiceDirectory =
+            Directory::Path(Path::Parse("/app/test/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE"));
+
+        Array<Kernel> kernels = Engine::DefaultKernels(spiceDirectory);
```
Comment (lucas-bremond) [inline test/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE.test.cpp:359]:
```suggestion
        const Array<Kernel> kernels = Engine::DefaultKernels(spiceDirectory);
```

File: test/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE.test.cpp [hunk 6]
```diff
@@ -339,3 +341,28 @@ TEST(OpenSpaceToolkit_Physics_Environment_Ephemerides_SPICE, AutomaticMode)
...
+    using ostk::core::ctnr::Array;
+    using ostk::core::fs::Path;
+    using ostk::core::fs::Directory;
+
+    using ostk::physics::env::ephem::spice::Engine;
+    using ostk::physics::env::ephem::spice::Kernel;
+
+    {
+
+        Directory spiceDirectory =
```
Comment (lucas-bremond) [inline test/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE.test.cpp:356]:
```suggestion
        const Directory spiceDirectory =
```

File: src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp [hunk 7]
```diff
@@ -354,9 +382,15 @@ void Engine::setup()
...
+        Directory kernelsDirectory = Manager::Get().getLocalRepository();
+        if (!kernelsDirectory.exists())
+        {
+            kernelsDirectory.create();
+        }
+
         // Load default kernels
 
-        for (const auto& kernel : Engine::DefaultKernels(Manager::Get().getLocalRepository()))
+        for (const auto& kernel : Engine::DefaultKernels(kernelsDirectory))
```
Comment (lucas-bremond) [inline src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp:393]:
```suggestion
        for (const auto& kernel : Engine::DefaultKernels(kernelDirectory))
```

File: src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp [hunk 8]
```diff
@@ -354,9 +382,15 @@ void Engine::setup()
 
     if (mode_ == Engine::Mode::Automatic)
     {
+        Directory kernelsDirectory = Manager::Get().getLocalRepository();
+        if (!kernelsDirectory.exists())
+        {
+            kernelsDirectory.create();
+        }
```
Comment (lucas-bremond) [inline src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp:389]:
```suggestion
        Directory kernelDirectory = Manager::Get().getLocalRepository();
        if (!kernelDirectory.exists())
        {
            kernelDirectory.create();
        }
```

File: src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp [hunk 9]
```diff
@@ -228,16 +231,41 @@ Engine::Mode Engine::DefaultMode()
...
+            const String filename = iter->path().filename().string();
+            if (std::filesystem::is_regular_file(*iter) && std::regex_match(filename, aRegex))
+            {
+                kernelPaths.add(Path::Parse(iter->path().string()));
+            }
+        }
+
+        if (kernelPaths.isEmpty())
+        {
+            return Manager::Get().fetchMatchingKernels(aRegex).accessFirst();
```
Comment (lucas-bremond) [inline src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp:256]: `accessFirst()` could fail is nothing is found: let's improve the error handling here?

File: src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp [hunk 10]
```diff
@@ -228,16 +231,41 @@ Engine::Mode Engine::DefaultMode()
...
 {
-    static const Array<Kernel> defaultKernels = {
+    // Use regex to pull Earth body shape, orientation and leap second kernels, as the file name is often updated.
+    using iterator = std::filesystem::directory_iterator;
 
-        Kernel::File(File::Path(aLocalRepository.getPath() + Path::Parse("naif0012.tls"))),  // Leap seconds
-        Kernel::File(File::Path(aLocalRepository.getPath() + Path::Parse("de430.bsp"))),     // Ephemeris
-        Kernel::File(File::Path(aLocalRepository.getPath() + Path::Parse("pck00011.tpc"))
-        ),  // System body shape and orientation constants
+    auto findKernel = [&aLocalRepository](const String aRegexString) -> Kernel
```
Comment (lucas-bremond) [inline src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp:237]: Why not making this a public method of the Manager class?

Comment (kyle-cochran) [reply src/OpenSpaceToolkit/Physics/Environment/Ephemerides/SPICE/Engine.cpp:237]: At this point it probably makes sense, huh? This quick fix MR has accidentally morphed into doing it the right way 😆

**Review by Derollez — APPROVED** [review]

**Pull request merged**